File: bindings/nodejs/tests/tuple_test.js
/*
 * tuple_test.js
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var fdb = require('../lib/fdb.js').apiVersion(500);
var fdbModule = require('../lib/fdbModule.js');

console.log(fdb.tuple.pack([-Math.pow(2,53)]));
console.log(fdb.tuple.pack([-Math.pow(2,53)+1]));

console.log(fdb.tuple.unpack(fdb.tuple.pack([-Math.pow(2,53)])));
console.log(fdb.tuple.unpack(fdb.tuple.pack([-Math.pow(2,53)+1])));

try {
	console.log(fdb.tuple.unpack(fdb.buffer.fromByteLiteral('\x0d\xdf\xff\xff\xff\xff\xff\xfe')));
}
catch(err) {
	console.log(err);
}

console.log(fdb.tuple.pack([0xff * 0xff]));
console.log(fdb.tuple.pack([0xffffffff + 100 ]));
console.log(fdb.buffer.printable(fdb.tuple.pack(['begin', [true, null, false], 'end'])))
console.log(fdb.tuple.unpack(fdb.buffer.fromByteLiteral('\x1a\xff\xff\xff\xff\xff\xff')));
console.log(fdb.tuple.unpack(fdb.tuple.pack(['TEST', 'herp', 1, -10, 393493, '\u0000abc', 0xffffffff + 100, true, false, [new Boolean(true), null, new Boolean(false), 0, 'asdf'], null])));
console.log(fdb.buffer.printable(fdb.tuple.pack([[[[['three']]], 'two'], 'one'])))
console.log(fdb.tuple.range(['TEST', 1]));
console.log(fdb.buffer.printable(fdb.tuple.pack([fdb.tuple.Float.fromBytes(new Buffer('402df854', 'hex')), fdb.tuple.Double.fromBytes(new Buffer('4005BF0A8B145769', 'hex')), new fdb.tuple.UUID(new Buffer('deadc0deba5eba115ca1ab1edeadc0de', 'hex'))])))
console.log(fdb.tuple.unpack(fdb.tuple.pack([fdb.tuple.Float.fromBytes(new Buffer('2734236f', 'hex'))])))

tuples = [
    [1,2],
    [1],
    [2],
    [true],
    [false],
    [1,true],
    [1,false],
    [1, []],
    [1, [null]],
    [1, [0]],
    [1, [1]],
    [1, [0,1,2]],
    [null],
    []
];
tuples.sort(fdb.tuple.compare);
console.log(tuples);

tuples = [
    [fdb.tuple.Float.fromBytes(new Buffer('2734236f', 'hex'))], // A really small value.
    [fdb.tuple.Float.fromBytes(new Buffer('80000000', 'hex'))], // -0.0
    [new fdb.tuple.Float(0.0)],
    [new fdb.tuple.Float(3.14)],
    [new fdb.tuple.Float(-3.14)],
    [new fdb.tuple.Float(2.7182818)],
    [new fdb.tuple.Float(-2.7182818)],
    [fdb.tuple.Float.fromBytes(new Buffer('7f800000', 'hex'))], // Infinity
    [fdb.tuple.Float.fromBytes(new Buffer('7fffffff', 'hex'))], // NaN
    [fdb.tuple.Float.fromBytes(new Buffer('ffffffff', 'hex'))], // -NaN
];
tuples.sort(fdb.tuple.compare);
console.log(tuples);

// Float overruns.
const floats = [ 2.037036e90, -2.037036e90, 4.9090935e-91, -4.9090935e-91, 2.345624805922133125e14, -2.345624805922133125e14 ];
for (var i = 0; i < floats.length; i++) {
    var f = floats[i];
    console.log(f + " -> " + fdb.tuple.Float.fromBytes((new fdb.tuple.Float(f)).toBytes()).value);
}

// Float type errors.
try {
    console.log((new fdb.tuple.Float("asdf")).toBytes());
} catch (e) {
    console.log("Caught!");
    console.log(e);
}

try {
    console.log(fdbModule.toFloat(3.14, 2.718));
} catch (e) {
    console.log("Caught!");
    console.log(e);
}

File: bindings/java/src/test/com/apple/foundationdb/test/TupleTest.java
/*
 * TupleTest.java
 *
 * This source file is part of the FoundationDB open source project
 *
 * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.apple.foundationdb.test;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

import com.apple.foundationdb.Database;
import com.apple.foundationdb.FDB;
import com.apple.foundationdb.TransactionContext;
import com.apple.foundationdb.subspace.Subspace;
import com.apple.foundationdb.tuple.Tuple;
import com.apple.foundationdb.tuple.Versionstamp;

public class TupleTest {
	private static final byte FF = (byte)0xff;

	public static void main(String[] args) throws NoSuchFieldException {
		final int reps = 1000;
		try {
			FDB fdb = FDB.selectAPIVersion(720);
			try(Database db = fdb.open()) {
				runTests(reps, db);
			}
		} catch(Throwable t) {
			t.printStackTrace();
		}
	}

	// Assumes API version < 520
	private static void incompleteVersionstamps300() {
		if(FDB.instance().getAPIVersion() >= 520) {
			throw new IllegalStateException("cannot run test with API version " + FDB.instance().getAPIVersion());
		}
		Tuple t1 = Tuple.from(Versionstamp.complete(new byte[]{FF, FF, FF, FF, FF, FF, FF, FF, FF, FF}), new byte[]{});
		Tuple t2 = Tuple.from(Versionstamp.incomplete());
		if(t1.equals(t2)) {
			throw new RuntimeException("tuples " + t1 + " and " + t2 + " compared equal");
		}
		byte[] bytes1 = t1.pack();
		byte[] bytes2 = t2.packWithVersionstamp();
		if(!Arrays.equals(bytes1, bytes2)) {
			throw new RuntimeException("tuples " + t1 + " and " + t2 + " did not have matching representations");
		}
		if(t1.equals(t2)) {
			throw new RuntimeException("tuples " + t1 + " and " + t2 + " compared equal with memoized packed representations");
		}

		// Make sure position information adjustment works.
		Tuple t3 = Tuple.from(Versionstamp.incomplete(1));
		if(t3.getPackedSize() != 1 + Versionstamp.LENGTH + Short.BYTES) {
			throw new RuntimeException("incomplete versionstamp has incorrect packed size " + t3.getPackedSize());
		}
		byte[] bytes3 = t3.packWithVersionstamp();
		if(ByteBuffer.wrap(bytes3, bytes3.length - Short.BYTES, Short.BYTES).order(ByteOrder.LITTLE_ENDIAN).getShort() != 1) {
			throw new RuntimeException("incomplete versionstamp has incorrect position");
		}
		if(!Tuple.fromBytes(bytes3, 0, bytes3.length - Short.BYTES).equals(Tuple.from(Versionstamp.incomplete(1)))) {
			throw new RuntimeException("unpacked bytes did not match");
		}
		Subspace subspace = new Subspace(Tuple.from("prefix"));
		byte[] bytes4 = subspace.packWithVersionstamp(t3);
		if(ByteBuffer.wrap(bytes4, bytes4.length - Short.BYTES, Short.BYTES).order(ByteOrder.LITTLE_ENDIAN).getShort() != 1 + subspace.getKey().length) {
			throw new RuntimeException("incomplete versionstamp has incorrect position with prefix");
		}
		if(!Tuple.fromBytes(bytes4, 0, bytes4.length - Short.BYTES).equals(Tuple.from("prefix", Versionstamp.incomplete(1)))) {
			throw new RuntimeException("unpacked bytes with subspace did not match");
		}

		// Make sure an offset > 0xFFFF throws an error.
		Tuple t4 = Tuple.from(Versionstamp.incomplete(2));
		byte[] bytes5 = t4.packWithVersionstamp(); // Get bytes memoized.
		if(ByteBuffer.wrap(bytes5, bytes5.length - Short.BYTES, Short.BYTES).order(ByteOrder.LITTLE_ENDIAN).getShort() != 1) {
			throw new RuntimeException("incomplete versionstamp has incorrect position with prefix");
		}
		byte[] bytes6 = t4.packWithVersionstamp(new byte[0xfffe]); // Offset is 0xffff
		if(!Arrays.equals(Arrays.copyOfRange(bytes5, 0, 1 + Versionstamp.LENGTH), Arrays.copyOfRange(bytes6, 0xfffe, 0xffff + Versionstamp.LENGTH))) {
			throw new RuntimeException("area before versionstamp offset did not match");
		}
		if((ByteBuffer.wrap(bytes6, bytes6.length - Short.BYTES, Short.BYTES).order(ByteOrder.LITTLE_ENDIAN).getShort() & 0xffff) != 0xffff) {
			throw new RuntimeException("incomplete versionstamp has incorrect position with prefix");
		}
		try {
			t4.packWithVersionstamp(new byte[0xffff]); // Offset is 0x10000
			throw new RuntimeException("able to pack versionstamp with offset that is too large");
		}
		catch(IllegalArgumentException e) {
			// eat
		}
		// Same as before, but packed representation is not memoized.
		try {
			Tuple.from(Versionstamp.incomplete(3)).packWithVersionstamp(new byte[0xffff]); // Offset is 0x10000
			throw new RuntimeException("able to pack versionstamp with offset that is too large");
		}
		catch(IllegalArgumentException e) {
			// eat
		}
	}


	private static void runTests(final int reps, TransactionContext db) {
		System.out.println("Running tests...");
		long start = System.currentTimeMillis();
		try {
			db.run(tr -> {
				Tuple t = new Tuple();
				t.add(100230045000L);
				t.add("Hello!");
				t.add("foo".getBytes());

				/*for(Map.Entry<byte[], byte[]> e : tr.getRange("vcount".getBytes(), "zz".getBytes())) {
					System.out.println("K: " + new String(e.getKey()) + ", V: " + new String(e.getValue()));
				}*/
				return null;
			});
		} catch (Throwable e) {
			e.printStackTrace();
		}
		long end = System.currentTimeMillis();

		double seconds = (end - start) / 1000.0;
		System.out.println(" Transactions:    " + reps);
		System.out.println(" Total Time:      " + seconds);
		System.out.println(" Gets+Sets / sec: " + reps / seconds);

		System.exit(0);
	}

	private TupleTest() {}
}
